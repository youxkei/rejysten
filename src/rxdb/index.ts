import { addRxPlugin } from "rxdb";
import { RxDBReplicationCouchDBPlugin } from "rxdb/plugins/replication-couchdb";
import { RxDBLeaderElectionPlugin } from "rxdb/plugins/leader-election";
import { RxDBUpdatePlugin } from "rxdb/plugins/update";
import { addPouchPlugin } from "rxdb/plugins/pouchdb";
import { RxDBMigrationPlugin } from "rxdb/plugins/migration";
import PouchDBAdapterIDB from "pouchdb-adapter-idb";
import PouchDBAdapterHTTP from "pouchdb-adapter-http";

export { useRxDatabase } from "@/rxdb/useRxDatabase";
export { useRxCollections } from "@/rxdb/useRxCollections";
export { useRxSubscribe } from "@/rxdb/useRxSubscribe";
export { useRxSync } from "@/rxdb/useRxSync";

addRxPlugin(RxDBLeaderElectionPlugin);
addRxPlugin(RxDBReplicationCouchDBPlugin);
addRxPlugin(RxDBUpdatePlugin);
addRxPlugin(RxDBMigrationPlugin);
addPouchPlugin(PouchDBAdapterIDB);
addPouchPlugin(PouchDBAdapterHTTP);
