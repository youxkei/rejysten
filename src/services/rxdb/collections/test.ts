import type { ListItem } from "@/services/rxdb/collections/listItem";

// export type ListItemFixture = [string, ListItemFixture[]] | string;
export type ListItemFixture = [string, ListItemFixture[]?, { text?: string }?];

export function makeListItems(parentId: string, updatedAt: number, fixtures: ListItemFixture[]): ListItem[] {
  const listItems = fixtures.map((fixture) => makeListItem(parentId, updatedAt, fixture));

  for (let i = 0; i < listItems.length; i++) {
    if (i - 1 >= 0) {
      listItems[i][0].prevId = listItems[i - 1][0].id;
    }

    if (i + 1 < listItems.length) {
      listItems[i][0].nextId = listItems[i + 1][0].id;
    }
  }

  return listItems.flat().sort((a, b) => a.id.localeCompare(b.id));
}

function makeListItem(parentId: string, updatedAt: number, [id, children, options]: ListItemFixture): ListItem[] {
  return [{ id, text: options?.text ?? id, prevId: "", nextId: "", parentId, updatedAt }, ...makeListItems(id, updatedAt, children ?? [])];
}
