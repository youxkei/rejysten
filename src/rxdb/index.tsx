import { JSX } from "solid-js";
import { RxDatabaseCreator, addRxPlugin } from "rxdb";
import { RxDBLeaderElectionPlugin } from "rxdb/plugins/leader-election";
import { RxDBUpdatePlugin } from "rxdb/plugins/update";
import { RxDBMigrationPlugin } from "rxdb/plugins/migration";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";

import { Provider as DatabaseProvider } from "@/rxdb/database";
import { Provider as CollectionsProvider } from "@/rxdb/collections";

addRxPlugin(RxDBLeaderElectionPlugin);
addRxPlugin(RxDBUpdatePlugin);
addRxPlugin(RxDBMigrationPlugin);

export function Provider(props: {
  databaseCreator?: RxDatabaseCreator;
  children: JSX.Element;
}) {
  return (
    <DatabaseProvider
      databaseCreator={
        props.databaseCreator ?? {
          name: "rejysten",
          storage: getRxStorageDexie(),
        }
      }
    >
      <CollectionsProvider>
        <>{props.children}</>
      </CollectionsProvider>
    </DatabaseProvider>
  );
}
