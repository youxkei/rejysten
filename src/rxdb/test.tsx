import type { Collections } from "@/rxdb/collections";
import type { JSX } from "solid-js";

import { indexedDB, IDBKeyRange } from "fake-indexeddb";
import { createRxDatabase, addRxPlugin } from "rxdb";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";
import { RxDBUpdatePlugin } from "rxdb/plugins/update";
import { Suspense } from "solid-js";

import { Provider as CollectionsProvider, collectionCreators } from "@/rxdb/collections";
import { Provider as DatabaseProvider } from "@/rxdb/database";

addRxPlugin(RxDBUpdatePlugin);

export function TestWithRxDB(props: { tid: string; children: JSX.Element }) {
  return (
    <Suspense fallback={props.tid}>
      <DatabaseProvider
        databaseCreator={{
          name: props.tid,
          storage: getRxStorageDexie({
            indexedDB: indexedDB,
            IDBKeyRange: IDBKeyRange,
          }),
          ignoreDuplicate: true,
        }}
      >
        <CollectionsProvider>{props.children}</CollectionsProvider>
      </DatabaseProvider>
    </Suspense>
  );
}

export async function createCollections(tid: string) {
  const database = await createRxDatabase<Collections>({
    name: tid,
    storage: getRxStorageDexie({
      indexedDB: indexedDB,
      IDBKeyRange: IDBKeyRange,
    }),
    ignoreDuplicate: true,
  });

  return database.addCollections(collectionCreators) as Promise<Collections>;
}
