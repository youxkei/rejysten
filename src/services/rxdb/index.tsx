import type { Collections } from "@/services/rxdb/collections";
import type { RxDatabase, RxDatabaseCreator } from "rxdb";
import type { DexieStorageInternals } from "rxdb/dist/types/types";
import type { JSXElement } from "solid-js";

import { addRxPlugin, createRxDatabase } from "rxdb";
import { RxDBMigrationPlugin } from "rxdb/plugins/migration";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";
import { Show } from "solid-js";
import { createContext, createResource, onCleanup, useContext } from "solid-js";

import { schema } from "@/services/rxdb/collections";

export type RxDBService = {
  collections: Collections;
};

const context = createContext<RxDBService>();

function createRxDBServiceSignal(databaseCreator$: () => RxDatabaseCreator<DexieStorageInternals>) {
  addRxPlugin(RxDBMigrationPlugin);

  const [database$] = createResource(databaseCreator$, async (databaseCreator) => {
    let database: RxDatabase<Collections> | undefined;

    onCleanup(() => {
      if (database) {
        database.destroy();
      }
    });

    database = await createRxDatabase<Collections>(databaseCreator);

    return database;
  });

  const [collections$] = createResource(database$, async (database) => {
    let collections: Collections | undefined;

    onCleanup(() => {
      if (collections) {
        for (const [_, collection] of Object.entries(collections)) {
          collection.destroy();
        }
      }
    });

    collections = await database.addCollections(schema as any); // TODO: somehow type check fails so far

    return collections;
  });

  return () => {
    const collections = collections$();
    if (!collections) return;

    return {
      collections,
    };
  };
}

export function RxDBServiceProvider(props: { children: JSXElement; databaseCreator?: RxDatabaseCreator<DexieStorageInternals> }) {
  const rxdbService$ = createRxDBServiceSignal(
    () =>
      props.databaseCreator ?? {
        name: "rejysten",
        storage: getRxStorageDexie(),
      }
  );

  return (
    <Show when={rxdbService$()}>{(rxdbService$) => <context.Provider value={rxdbService$()}>{props.children}</context.Provider>}</Show>
  );
}

export function useRxDBService() {
  const service = useContext(context);
  if (!service) throw new Error("useRxDBService must be used within RxDBServiceProvider");

  return service;
}
