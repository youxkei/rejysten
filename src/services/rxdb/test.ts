import type { Collections } from "@/services/rxdb/collections";

import { indexedDB, IDBKeyRange } from "fake-indexeddb";
import { createRxDatabase } from "rxdb";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";

import { schema } from "@/services/rxdb/collections";

export function createDatabaseCreatorForTest(tid: string) {
  return {
    name: tid,
    storage: getRxStorageDexie({ indexedDB, IDBKeyRange }),
    ignoreDuplicate: true,
  };
}

export async function createCollectionsForTest(tid: string) {
  const database = await createRxDatabase<Collections>({
    name: tid,
    storage: getRxStorageDexie({ indexedDB, IDBKeyRange }),
    ignoreDuplicate: true,
  });

  return database.addCollections(schema) as Promise<Collections>;
}
