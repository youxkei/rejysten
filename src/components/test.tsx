import type { JSXElement } from "solid-js";

import { Suspense } from "solid-js";

import { RxDBServiceProvider } from "@/services/rxdb";
import { createDatabaseCreatorForTest } from "@/services/rxdb/test";

export function TestWithRxDBService(props: { tid: string; children: JSXElement }) {
  return (
    <Suspense fallback={props.tid}>
      <RxDBServiceProvider databaseCreator={createDatabaseCreatorForTest(props.tid)}>
        {props.children}
      </RxDBServiceProvider>
    </Suspense>
  );
}

export { createCollectionsForTest } from "@/services/rxdb/test";
