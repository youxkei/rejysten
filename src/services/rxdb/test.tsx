import type { Collections } from "@/services/rxdb/collections";
import type { JSXElement } from "solid-js";

import { indexedDB, IDBKeyRange } from "fake-indexeddb";
import { createRxDatabase } from "rxdb";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";

import { RxDBServiceProvider } from "@/services/rxdb";
import { schema } from "@/services/rxdb/collections";

type Wait = { promise: Promise<Wait> };

export function createDatabaseCreatorForTest(tid: string) {
  return {
    name: tid,
    storage: getRxStorageDexie({ indexedDB, IDBKeyRange }),
    ignoreDuplicate: true,
  };
}

export async function createCollectionsForTest(tid: string) {
  const database = await createRxDatabase<Collections>(createDatabaseCreatorForTest(tid));

  return database.addCollections(schema as any) as Promise<Collections>; // TODO: somehow type check fails so far
}

export function RxDBServiceProviderForTest(props: { tid: string; children: JSXElement }) {
  return <RxDBServiceProvider databaseCreator={createDatabaseCreatorForTest(props.tid)}>{props.children}</RxDBServiceProvider>;
}
