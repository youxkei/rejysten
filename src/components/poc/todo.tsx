import { Ulid } from "id128";
import { For } from "solid-js";

import { useCollections, useSubscribe, useSubscribeAll } from "@/rxdb";

export function Todo() {
  const collections = useCollections();

  const originalTodos = useSubscribeAll(() => collections()?.todos.find());
  const todos = () => {
    const todos = originalTodos();

    return todos;
  };

  const editor = useSubscribe(() => collections()?.editors.findOne("const"));

  const text = () => editor()?.text ?? "";

  const onInput = (event: { currentTarget: HTMLInputElement }) => {
    collections()?.editors.upsert({
      id: "const",
      text: event.currentTarget.value,
      updatedAt: Date.now(),
    });
  };

  const onClick = () => {
    const id = Ulid.generate();

    collections()?.todos.insert({
      id: id.toCanonical(),
      text: text(),
      updatedAt: id.time.getTime(),
    });

    collections()?.editors.upsert({
      id: "const",
      text: "",
      updatedAt: Date.now(),
    });
  };

  return (
    <>
      <ul>
        <For each={todos()}>{(todo, _) => <li>{todo.text}</li>}</For>
      </ul>
      <p>{text()}</p>
      <input value={text()} onInput={onInput} />
      <button onClick={onClick}>add</button>
    </>
  );
}

import {
  render,
  waitForElementToBeRemoved,
  findByText,
  queryByText,
} from "@solidjs/testing-library";
import userEvent from "@testing-library/user-event";

import { TestWithRxDB, createCollections } from "@/rxdb/test";

if (import.meta.vitest) {
  test("renders", async (ctx) => {
    const tid = ctx.meta.id;
    let collections = await createCollections(tid);

    await collections.todos.bulkUpsert([
      { id: "001", text: "foo", updatedAt: 1 },
      { id: "002", text: "bar", updatedAt: 2 },
    ]);

    const { container, unmount } = render(() => (
      <TestWithRxDB tid={tid}>
        <Todo />
      </TestWithRxDB>
    ));

    await waitForElementToBeRemoved(() => queryByText(container, tid));
    ctx.expect(container).toMatchSnapshot();

    unmount();
  });

  test("add todos", async (ctx) => {
    const tid = ctx.meta.id;
    const collections = await createCollections(tid);
    const user = userEvent.setup();

    await collections.todos.bulkUpsert([
      { id: "001", text: "foo", updatedAt: 1 },
      { id: "002", text: "bar", updatedAt: 2 },
    ]);

    const { container, unmount } = render(() => (
      <TestWithRxDB tid={tid}>
        <Todo />
      </TestWithRxDB>
    ));

    await waitForElementToBeRemoved(() => queryByText(container, tid));
    ctx.expect(container).toMatchSnapshot();

    const input = container.querySelector("input")!;
    const button = container.querySelector("button")!;

    await user.click(input);
    await user.keyboard("baz");
    await findByText(container, "baz");
    ctx.expect(container).toMatchSnapshot();

    await user.click(button);
    await waitForElementToBeRemoved(() =>
      queryByText(container.querySelector("p")!, "baz")
    );
    await findByText(container.querySelector("ul")!, "baz");
    ctx.expect(container).toMatchSnapshot();

    unmount();
  });
}
