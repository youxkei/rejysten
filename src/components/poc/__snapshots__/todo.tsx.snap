// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add todos 1`] = `
<div>
  <div>
    <ul>
      <li>
        foo
      </li>
      <li>
        bar
      </li>
    </ul>
    <p />
    <input />
    <button>
      add
    </button>
  </div>
</div>
`;

exports[`add todos 2`] = `
<div>
  <div>
    <ul>
      <li>
        foo
      </li>
      <li>
        bar
      </li>
    </ul>
    <p>
      the quick brown fox jumps over the lazy dog
    </p>
    <input />
    <button>
      add
    </button>
  </div>
</div>
`;

exports[`add todos 3`] = `
<div>
  <div>
    <ul>
      <li>
        foo
      </li>
      <li>
        bar
      </li>
      <li>
        the quick brown fox jumps over the lazy dog
      </li>
    </ul>
    <p>
      
    </p>
    <input />
    <button>
      add
    </button>
  </div>
</div>
`;

exports[`renders 1`] = `
<div>
  <div>
    <ul>
      <li>
        foo
      </li>
      <li>
        bar
      </li>
    </ul>
    <p />
    <input />
    <button>
      add
    </button>
  </div>
</div>
`;
